File: Codeforces Round #821 (Div. 2)/F.py
from functools import lru_cache as cache
from collections import Counter, defaultdict
import math
import bisect
from heapq import *
# from sortedcontainers import SortedSet, SortedList, SortedDict

import sys
def get_int(): return int(input())
def get_ints(): return map(int, sys.stdin.readline().strip().split())
def get_list(): return list(map(int, sys.stdin.readline().strip().split()))
def get_string(): return sys.stdin.readline().strip()

def main():
    testcases = get_int()
    for tc in range(testcases):
        n, x, y = get_ints()
        a = get_string()
        b = get_string()
        print(solve(n, x, y, a, b))

def solve(n, x, y, a, b):
    diffs = []
    for i in range(n):
        if a[i] != b[i]:
            diffs.append(i)
    if len(diffs) % 2 == 1:
        return -1
    if x >= y:
        return solve1(x, y, diffs)
    else:
        return solve2(x, y, diffs)

def solve1(x, y, diffs):
    if len(diffs) == 2 and diffs[0] + 1 == diffs[1]:
        return min(x, y*2)
    return y * len(diffs) // 2
        
def solve2(x, y, diffs):
    m = len(diffs)
    @cache(None)
    def dfs(i, drop):
        if i >= m:
            if drop == 0:
                return 0
            return math.inf
        result = dfs(i+1, drop+1)
        if drop:
            result = min(result, dfs(i+1, drop-1) + y)
        if i < m-1:
            result = min(result, dfs(i+2, drop) + min(y, x * (diffs[i+1] - diffs[i])))
        return result
    return dfs(0, 0)


if __name__ == "__main__":
    main()

File: Codeforces Round #821 (Div. 2)/F.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var reader *bufio.Reader = bufio.NewReader(os.Stdin)
var writer *bufio.Writer = bufio.NewWriter(os.Stdout)

func printf(f string, a ...interface{}) { fmt.Fprintf(writer, f, a...) }
func print(a ...interface{})            { fmt.Println(a...) }
func scanf(f string, a ...interface{})  { fmt.Fscanf(reader, f+"\n", a...) }

func main() {
	defer writer.Flush()

	var testcases int64
	scanf("%d", &testcases)
  var i int64
	for i = 0; i < testcases; i++ {
		var n, x, y int64
		scanf("%d %d %d", &n, &x, &y)
    var a, b string
    scanf("%s", &a)
    scanf("%s", &b)

		print(solve(n, x, y, a, b))
	}
}

func max(nums ...int64) int64 {
  result := nums[0]
	for _, num := range nums {
		if num > result {
			result = num
		}
	}
	return result
}

func min(nums ...int64) int64 {
  result := nums[0]
  for _, num := range nums {
    if num < result {
      result = num
    }
  }
  return result
}

func solve(n, x, y int64, a, b string) int64 {
  diffs := make([]int64, 0, n)
  var i int64
  for i = 0; i < n; i++ {
    if a[i] != b[i] {
      diffs = append(diffs, i)
    }
  }

  m := int64(len(diffs))

  if m % 2 == 1 {
    return -1
  }

  if x >= y {
    if m == 2 && diffs[0] + 1 == diffs[1] {
      return min(x, y*2)
    }
    return y * m / 2
  }

  dp := make([][]int64, m)
  for i = 0; i < m; i++ {
    dp[i] = make([]int64, m)
    var j int64
    for j = 0; j < m; j++ {
      dp[i][j] = -1
    }
  }
  var dfs func(i, drop int64) int64
  dfs = func(i, drop int64) int64 {
    if i >= m {
      if drop == 0 {
        return 0
      }
      return math.MaxInt64
    }
    if dp[i][drop] != -1 {
      return dp[i][drop]
    }
    result := dfs(i+1, drop+1)
    if drop > 0 {
      this := dfs(i+1, drop-1)
      if this != math.MaxInt64 {
        result = min(result, this + y)
      }
    }
    if i < m-1 {
      this := dfs(i+2, drop)
      if this != math.MaxInt64 {
        result = min(result, this + min(y, x * (diffs[i+1] - diffs[i])))
      }
    }
    dp[i][drop] = result
    return result
  }
  res := dfs(0, 0)
  return res
}

